libraries/standard/retry-utils/source/include/retry_utils.h

```diff
@@ -0,0 +1,174 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file retry_utils.h
+ * @brief Declaration of retry utility functions and constants for exponential backoff with
+ * jitter strategy of retry attempts.
+ * This library represents the "Full Jitter" backoff strategy explained in the
+ * following document.
+ * https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
+ *
+ */
+
+#ifndef RETRY_UTILS_H_
+#define RETRY_UTILS_H_
+
+/* Standard include. */
+#include <stdint.h>
+
+/**
+ * @page retryutils_page Retry Utilities
+ * @brief An abstraction of utilities for retrying with exponential back off and
+ * jitter.
+ *
+ * @section retryutils_overview Overview
+ * The retry utilities are a set of APIs that aid in retrying with exponential
+ * backoff and jitter. Exponential backoff with jitter is strongly recommended
+ * for retrying failed actions over the network with servers. Please see
+ * https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/ for
+ * more information about the benefits with AWS.
+ *
+ * Exponential backoff with jitter is typically used when retrying a failed
+ * connection to the server. In an environment with poor connectivity, a client
+ * can get disconnected at any time. A backoff strategy helps the client to
+ * conserve battery by not repeatedly attempting reconnections when they are
+ * unlikely to succeed.
+ *
+ * Before retrying the failed communication to the server, there is a delay period.
+ * In this delay period, the task that is retrying must sleep for some random amount
+ * of seconds between 0 and the lesser of the backoff window (related to the retry attempt)
+ * and a predefined maximum delay value. The backoff window is doubled with each retry
+ * attempt until the maximum delay value is reached.<br>
+ *
+ * > sleep_seconds = random_between( 0, min( 2<sup>attempts_count</sup> * base_seconds, maximum_seconds ) )
```

Comment:
I think sleep_seconds may not be accurate as we are using ms. Similar comment for base_seconds and maximum_seconds

Reply:
Updated to milliseconds